[serve][docs] Document lightweight config updates (#27706)
ray-project/ray: doc/source/serve/production.md

## Updating Your Serve Application in Production

> You can also update your Serve applications once they're in production. You can update the settings in your config file and redeploy it using the `serve deploy` command. You can also add new deployment settings or remove old deployment settings from the config. This is because `serve deploy` is **idempotent**. Your Serve application's will match the one specified in the latest config you deployed– regardless of what config files you deployed before that.

Reviewer: ```suggestion
You can also update your Serve applications once they're in production. You can update the settings in your config file and redeploy it using the `serve deploy` command. You can also add new deployment settings or remove old deployment settings from the config. This is because `serve deploy` is **idempotent**. Your Serve application's config always matches the latest config you deployed successfully– regardless of what config files you deployed before that.
```

> You can also update your Serve applications once they're in production. You can update the settings in your config file and redeploy it using the `serve deploy` command. You can also add new deployment settings or remove old deployment settings from the config. This is because `serve deploy` is **idempotent**. Your Serve application's config always matches the latest config you deployed successfully– regardless of what config files you deployed before that.

Reviewer 2: typo? `successfully–`

> You can also update your Serve applications once they're in production. You can update the settings in your config file and redeploy it using the `serve deploy` command. You can also add new deployment settings or remove old deployment settings from the config. This is because `serve deploy` is **idempotent**. Your Serve application's config always matches the latest config you deployed successfully– regardless of what config files you deployed before that.

Reviewer 3: The first three sentences begin with the same subject "You". Vary it slightly. One way to write this to enumerate actions you can. perform to an existing application in production. One suggestion is:

"A Serve application configuration in production can be modified in a number ways:

- by updating the settings in its config file and redeploying with `serve deploy` CLI
- by adding new or removing old deployment settings from the config file 

All these updates are possible because `serve deploy` is **idempotent**, meaning your Serve application's config always matches (or honors) the latest config in the file.

### Lightweight Config Updates

> Lightweight config updates don't need to tear down the running deployments, meaning there's less downtime as the deployments update. Changing `num_replicas`, `autoscaling_config`, and/or `user_config` in an entry of the `deployments` section of the config file is considered a lightweight config update, and won't tear down the replicas for that deployment.

Reviewer: ```suggestion
Lightweight config updates modify running deployment replicas without tearing them down and restarting them, so there's less downtime as the deployments update. Changing `num_replicas`, `autoscaling_config`, and/or `user_config` in an entry of the `deployments` section of the config file is considered a lightweight config update, and won't tear down the replicas for that deployment.
```

> Lightweight config updates modify running deployment replicas without tearing them down and restarting them, so there's less downtime as the deployments update. Changing `num_replicas`, `autoscaling_config`, and/or `user_config` in an entry of the `deployments` section of the config file is considered a lightweight config update, and won't tear down the replicas for that deployment.

Reviewer 2: Hi @shrekris-anyscale , is it functioning already? or we still rely on `version` to keep the running replica alive?

Author: Yes, this was added in https://github.com/ray-project/ray/pull/27444

> 

Reviewer 3: Lightweight config updates modify running deployment replicas without tearing them down or restarting them, so there's less downtime on deployments update. A config file entry under `deployments`, such as `num_replicas`, `autoscaling_config`, or `user_config`, is considered a lightweight config update, therefore it won't tear down the replicas. Nor does it require a redeployment.

The price has updated! The same request now returns `10` instead of `6`, reflecting the new price.

### Code Updates

> 

Reviewer: Can we add a note somewhere that changing the top-level `import_path` or `runtime_env` will tear down all the deployments and restart them?

The price has updated! The same request now returns `10` instead of `6`, reflecting the new price.

### Code Updates

Similarly, you can update any other setting in any deployment in the config file. If a deployment setting other than `num_replicas`, `autoscaling_config`, or `user_config` is changed, it is considered a code update and the deployment replicas will be restarted. Note that the following are all considered "changes", and (if considered a code update) will trigger tear down of replicas:
* Changing an existing setting.
* Adding an override setting that was previously not present in the config file.
* Removing a setting from the config file.

> 

Reviewer 3: If a deployment setting other than `num_replicas`, `autoscaling_config`, or `user_config` is changed, it is considered a code update, and the deployment replicas will be restarted. Note that the following modifications are all considered "changes," and (considered as a code update) will trigger tear down of replicas:
* changing an existing setting
* adding an override setting that was previously not present in the config file
* removing a setting from the config file
